File: cpp_solutions/Array/3Sum.cpp
/*
 * Given an array S of n integers, are there elements a, b, c in S such that a + b + c = 0? 
 * Find all unique triplets in the array which gives the sum of zero.
 *
 * */

// Notes: two pointers

#include<vector>
#include<algorithm>

using namespace std;

vector<vector<int>> threeSum(vector<int>& nums)
{
	vector<vector<int>> rst;

	// sort the array
	sort(nums.begin(), nums.end());

	for(int i=0; i < nums.size(); i++){
		if(i > 0 && nums[i] == nums[i-1])
		{
			continue;
		}

		// convert this problem to 2Sum
		int start = i+1, end = nums.size() - 1;
		int target = -nums[i];

		while(start < end){
			if(start > i+1 && nums[start] == nums[start-1]){
				start++;
				continue;
			}

			if(nums[start] + nums[end] < target)
			{
				start++ ;
			}
			else if(nums[start] + nums[end] > target)
			{
				end--;
			}
			else
			{
				vector<int> temp;
				temp.push_back(nums[i]);
				temp.push_back(nums[start]);
				temp.push_back(nums[end]);
				rst.push_back(temp);
				start++;
			}
		}
	}	
	return rst;
}

File: python_solutions/Array/3sum.py
class Solution:
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        length = len(nums)
        ret = []
        for i in range(length-2):
            if i and nums[i] == nums[i-1]:
                continue
            left, right = i+1, length-1
            target = -1*nums[i]
            while left<right:  
                if nums[left]+nums[right] == target:
                    ret.append([nums[i], nums[left], nums[right]])
                    left+=1
                    right-=1
                    while left < right and nums[left] == nums[left - 1]:
                        left+=1
                    while left < right and nums[right] == nums[right+1]:
                        right-=1
                elif nums[left]+nums[right] < target:
                     left+=1
                else:
                     right-=1
        return ret